Root package.json break publishing

### Describe the problem

Publishing fail when a project (e.g. [Maudit](https://github.com/bruits/maudit/)) have a root `package.json` without `version` (even with `"type": "module", "private": true`). Reported by @Princesseuh 

```
Error: Failed to execute sampo publish: sampo publish failed: Publish error: Manifest /home/runner/work/maudit/maudit/package.json is missing a non-empty 'version' field
Error: Process completed with exit code 1.
```

### What's the expected result?

Should probably ignore workspace root.

--- a/.sampo/changesets/regal-guardian-lemminkainen.md
+++ b/.sampo/changesets/regal-guardian-lemminkainen.md
@@ -1,0 +1,7 @@
+---
+cargo/sampo: patch
+cargo/sampo-core: patch
+cargo/sampo-github-action: patch
+---
+
+Npm packages marked `private: true` no longer block Sampo publish when their manifest omits `version`.

--- a/crates/sampo-core/src/adapters/npm.rs
+++ b/crates/sampo-core/src/adapters/npm.rs
@@ -29,7 +29,7 @@ struct NpmPublishConfig {
 struct NpmManifestInfo {
     name: String,
     #[allow(dead_code)]
-    version: String,
+    version: Option<String>,
     private: bool,
     package_manager: Option<String>,
     publish_config: NpmPublishConfig,
@@ -215,17 +215,19 @@ fn parse_manifest_info(manifest_path: &Path, manifest: &JsonValue) -> Result<Npm
         .and_then(JsonValue::as_str)
         .map(str::trim)
         .filter(|s| !s.is_empty())
-        .ok_or_else(|| {
-            SampoError::Publish(format!(
-                "Manifest {} is missing a non-empty 'version' field",
-                manifest_path.display()
-            ))
-        })?;
+        .map(|s| s.to_string());
 
     let private = manifest
         .get("private")
         .and_then(JsonValue::as_bool)
         .unwrap_or(false);
+
+    if !private && version.is_none() {
+        return Err(SampoError::Publish(format!(
+            "Manifest {} is missing a non-empty 'version' field",
+            manifest_path.display()
+        )));
+    }
 
     let package_manager = manifest
         .get("packageManager")
@@ -263,7 +265,7 @@ fn parse_manifest_info(manifest_path: &Path, manifest: &JsonValue) -> Result<Npm
 
     Ok(NpmManifestInfo {
         name: name.to_string(),
-        version: version.to_string(),
+        version,
         private,
         package_manager,
         publish_config,

--- a/crates/sampo-core/src/adapters/npm/npm_tests.rs
+++ b/crates/sampo-core/src/adapters/npm/npm_tests.rs
@@ -176,13 +176,33 @@ fn parse_manifest_reads_publish_config() {
     });
     let info = super::parse_manifest_info(Path::new("/repo/package.json"), &manifest).unwrap();
     assert_eq!(info.name, "pkg");
-    assert_eq!(info.version, "1.2.3");
+    assert_eq!(info.version.as_deref(), Some("1.2.3"));
     assert_eq!(
         info.publish_config.registry.as_deref(),
         Some("https://registry.example.com/")
     );
     assert_eq!(info.publish_config.access.as_deref(), Some("restricted"));
     assert_eq!(info.publish_config.tag.as_deref(), Some("beta"));
+}
+
+#[test]
+fn parse_manifest_allows_private_package_without_version() {
+    let manifest: serde_json::Value =
+        serde_json::json!({"name": "pkg", "private": true, "type": "module"});
+    let info = super::parse_manifest_info(Path::new("/repo/package.json"), &manifest).unwrap();
+    assert!(info.version.is_none());
+    assert!(info.private);
+}
+
+#[test]
+fn parse_manifest_requires_version_for_publishable_package() {
+    let manifest: serde_json::Value = serde_json::json!({"name": "pkg"});
+    let err = super::parse_manifest_info(Path::new("/repo/package.json"), &manifest).unwrap_err();
+    let msg = err.to_string();
+    assert!(
+        msg.contains("missing a non-empty 'version' field"),
+        "unexpected error message: {msg}"
+    );
 }
 
 #[test]

